### PR TITLE
Add unit tests to data cleaning and operation scripts

### PR DESCRIPTION
**Pull request by Tony911029**
1. Add unit tests to data cleaning and operation scripts
2. Add Pydantic model (maybe we could make MealRecord universal like an object that we can use everywhere to validate data like typescript interface)
3. Some minor refactoring. Move plotting to `ploy.py`

### CODE REVIEW

**Comment by Tony911029** [comment]
Oops. Merge conflict. Will resolve that later today

File: 0_meal_identification/meal_identification/meal_identification/datasets/PydanticModels.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        valid_types = {'ANNOUNCE_MEAL', 'DOSE_INSULIN', '', '0', }
+        if v not in valid_types:
+            raise ValueError(f'msg_type must be one of {valid_types}')
+        return v
+
+    @validator('bgl')
+    def validate_bgl(cls, v):
+        if v <= 0:
+            raise ValueError('Blood glucose level must be positive')
+        if v > 1000:
```
Comment (RobotPsychologist) [inline 0_meal_identification/meal_identification/meal_identification/datasets/PydanticModels.py:45]: We can lower this threshold to be 600

File: 0_meal_identification/meal_identification/meal_identification/datasets/PydanticModels.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    @validator('msg_type')
+    def validate_msg_type(cls, v):
+        valid_types = {'ANNOUNCE_MEAL', 'DOSE_INSULIN', '', '0', }
+        if v not in valid_types:
+            raise ValueError(f'msg_type must be one of {valid_types}')
+        return v
+
+    @validator('bgl')
+    def validate_bgl(cls, v):
+        if v <= 0:
```
Comment (RobotPsychologist) [inline 0_meal_identification/meal_identification/meal_identification/datasets/PydanticModels.py:43]: Switch to 20

File: 0_meal_identification/meal_identification/meal_identification/datasets/PydanticModels.py
Comment (RobotPsychologist) [inline 0_meal_identification/meal_identification/meal_identification/datasets/PydanticModels.py:1]: please change file name to convention.

File: 0_meal_identification/meal_identification/meal_identification/datasets/dataset_cleaner.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    pd.DataFrame
+        The processed DataFrame with meal overlaps handled.
+    """
+    announce_meal_mask = patient_df['msg_type'] == 'ANNOUNCE_MEAL'
+    announce_meal_indices = patient_df[announce_meal_mask].index
+
+    for idx in announce_meal_indices:
+        # Skip meals below the carbohydrate threshold
+        if patient_df.at[idx, 'food_g'] <= min_carbs:
+            continue
```
Comment (RobotPsychologist) [inline 0_meal_identification/meal_identification/meal_identification/datasets/dataset_cleaner.py:27]: Change msg_type to "LOW_CARB_MEAL"

File: 0_meal_identification/meal_identification/meal_identification/datasets/dataset_operations.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+    # Resample the data
+    data_resampled = data.resample(resample_rule).first()
+
+    # Separate meal announcements and non-meal data
+    meal_announcements = data_resampled[data_resampled['msg_type'] == 'ANNOUNCE_MEAL'].copy()
+    non_meals = data_resampled[data_resampled['msg_type'] != 'ANNOUNCE_MEAL'].copy()
+
+    # Resample meal announcements separately
+    meal_announcements = meal_announcements.resample('5min').first()
```
Comment (RobotPsychologist) [inline 0_meal_identification/meal_identification/meal_identification/datasets/dataset_operations.py:184]: Switch to not be hard coded 5min to sampling window parameter.

File: 0_meal_identification/meal_identification/meal_identification/datasets/dataset_operations.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    # Convert 'date' to datetime if not already
+    if not pd.api.types.is_datetime64_any_dtype(data['date']):
+        data['date'] = pd.to_datetime(data['date'], errors='coerce')
+        data = data.dropna(subset=['date'])  # Drop rows where 'date' couldn't be parsed
+
+    # Set 'date' as the index without squeezing
+    data = data.set_index('date').sort_index()
+
+    # Define resample rule based on the provided timedelta
+    resample_rule = f'{int(coerse_time_interval.total_seconds() // 60)}min'  # e.g., '5min' for 5 minutes
```
Comment (RobotPsychologist) [inline 0_meal_identification/meal_identification/meal_identification/datasets/dataset_operations.py:174]: Get rid of this and follow https://pandas.pydata.org/docs/reference/api/pandas.Timedelta.html
for options. Let pandas handle the timing.

File: 0_meal_identification/meal_identification/meal_identification/datasets/dataset_operations.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+    # Resample the data
+    data_resampled = data.resample(resample_rule).first()
```
Comment (RobotPsychologist) [inline 0_meal_identification/meal_identification/meal_identification/datasets/dataset_operations.py:177]: remove this line.

Comment (RobotPsychologist) [reply 0_meal_identification/meal_identification/meal_identification/datasets/dataset_operations.py:177]: We don't want to erase meal overlaps.

File: 0_meal_identification/meal_identification/meal_identification/datasets/dataset_operations.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+    # Separate meal announcements and non-meal data
+    meal_announcements = data_resampled[data_resampled['msg_type'] == 'ANNOUNCE_MEAL'].copy()
+    non_meals = data_resampled[data_resampled['msg_type'] != 'ANNOUNCE_MEAL'].copy()
+
+    # Resample meal announcements separately
+    meal_announcements = meal_announcements.resample('5min').first()
+    non_meals = non_meals.resample('5min').first()
```
Comment (RobotPsychologist) [inline 0_meal_identification/meal_identification/meal_identification/datasets/dataset_operations.py:185]: switch from 5min to parameter.

File: 0_meal_identification/meal_identification/meal_identification/datasets/dataset_operations.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+    # Join the two DataFrames
+    data_resampled = non_meals.join(meal_announcements, how='left', rsuffix='_meal')
+
+    # Combine the columns
+    for col in ['bgl', 'msg_type', 'food_g']:
+        meal_col = f"{col}_meal"
+        if meal_col in data_resampled.columns:
+            data_resampled[col] = data_resampled[col + '_meal'].combine_first(data_resampled[col])
+            data_resampled.drop(columns=[meal_col], inplace=True)
```
Comment (RobotPsychologist) [inline 0_meal_identification/meal_identification/meal_identification/datasets/dataset_operations.py:195]: Delete the drop line.

File: 0_meal_identification/meal_identification/meal_identification/datasets/dataset_operations.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+    data_resampled['food_g_keep'] = data_resampled.get('food_g_meal', 0)
+
+    # Identify columns that end with '_meal'
+    columns_to_drop = data_resampled.filter(regex='_meal$').columns
+
+    # Drop the identified columns
+    data_resampled = data_resampled.drop(columns=columns_to_drop)
+
+    # At this point, 'date' is still the index. Do NOT reset the index.
+    data_resampled['date'] = data_resampled.index
```
Comment (RobotPsychologist) [inline 0_meal_identification/meal_identification/meal_identification/datasets/dataset_operations.py:207]: This line is probably redundant. Check to make sure.

**Review by RobotPsychologist — APPROVED** [review]
Please fix minor bugs identified from other comments.

**Pull request merged**